core/src/Kokkos_Swap.hpp

```diff
@@ -0,0 +1,67 @@
+//@HEADER
+// ************************************************************************
+//
+//                        Kokkos v. 4.0
+//       Copyright (2022) National Technology & Engineering
+//               Solutions of Sandia, LLC (NTESS).
+//
+// Under the terms of Contract DE-NA0003525 with NTESS,
+// the U.S. Government retains certain rights in this software.
+//
+// Part of Kokkos, under the Apache License v2.0 with LLVM Exceptions.
+// See https://kokkos.org/LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//@HEADER
+
+#ifndef KOKKOS_SWAP_HPP
+#define KOKKOS_SWAP_HPP
+
+#include <Kokkos_Macros.hpp>
+
+#include <cstddef>
+#include <type_traits>
+#include <utility>
+
+namespace Kokkos {
+
+template <class T>
+KOKKOS_FUNCTION constexpr std::enable_if_t<std::is_move_constructible_v<T> &&
+                                           std::is_move_assignable_v<T>>
+swap(T& a, T& b) noexcept(std::is_nothrow_move_constructible_v<T>&&
```

Comment:
I am not sure providing a custom `Kokkos::swap` implementation for *any* object type is a good idea.

For example, any host compiler I know supports `std::swap` (and `Kokkos` is on `c++17` so `std::swap` is OK on host code whatsoever).

For device compilers, some are already supporting it. Those I tested:
- `nvcc` (`Build cuda_12.2.r12.2/compiler.33191640_0`)
- `hipcc` (`HIP version: 5.7.31921-d1770ee1b`)

Reply:
As far as I can tell it is not supported by NVCC https://godbolt.org/z/9s3Px96rM

Reply:
Sorry, it works on our side for 2 reasons:
1. We are on `c++20`, so `std::swap` is `constexpr`.
2. We compile with the flag `--expt-relaxed-constexpr` when using `nvcc`

So yes, it seems you'll need to bother with re-implementing the wheel once again :wheel_of_dharma: